### PR TITLE
wallet: allow zero-fee fundrawtransaction/walletcreatefundedpsbt and other fixes

### PR DESCRIPTION
**Pull request by jonatack**
- Fixes https://github.com/bitcoin/bitcoin/pull/20305/files#r525406176. A check to raise an error on zero-fee txns was mistakenly extended in a0d4957 from the bumpfee and send{toaddress, many} RPCs to also include fundrawtransaction and walletcreatefundedpsbt. This commit re-overrides zero fee rate checking for these two RPCs, not only for the feeRate (BTC/kvB) arg to return to previous behavior, but also for the new fee_rate (sat/vB) arg. Negative fee rates will still raise "amount out of range" by the MoneyRange check in src/bitcoin-tx.cpp::AmountFromValue.

- Updates a wallet bumpfee test from feeRate (BTC/kvB) to fee_rate (sat/vB)

- Fixes https://github.com/bitcoin/bitcoin/pull/20305/files#r525405363 to use the correct incremental fee rate constant in the bumpfee help  (thanks Marco Falke for the catch) and rectifies "1.000 sat/vB sat/vB" in the help to "1.000 sat/vB"

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -219,14 +219,8 @@ static void SetFeeEstimateMode(const CWallet* pwallet, CCoinControl& cc, const U
...
-        CFeeRate fee_rate_in_sat_vb{CFeeRate(AmountFromValue(fee_rate), COIN)};
-        if (override_min_fee) {
-            if (fee_rate_in_sat_vb <= CFeeRate(0)) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid fee_rate %s (must be greater than 0)", fee_rate_in_sat_vb.ToString(FeeEstimateMode::SAT_VB)));
-            }
-            cc.fOverrideFeeRate = true;
-        }
-        cc.m_feerate = fee_rate_in_sat_vb;
+        cc.m_feerate = CFeeRate(AmountFromValue(fee_rate), COIN);
+        if (override_min_fee) cc.fOverrideFeeRate = true;
```
Comment (Xekyo) [inline src/wallet/rpcwallet.cpp:222]: I see that we're no longer checking for smaller than zero. Are we just relying on that being invalid for a transaction in the first place?

**Review by murchandamus — APPROVED** [review]
utACK

**Comment by MarcoFalke** [comment]
in commit 3856f0fdc41fa72e8123f5ba64077d62e8163b4f: Could add test for

* 0-fee for `send*` rpcs, and
* negative fee for all rpcs?

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (jonatack) [reply src/wallet/rpcwallet.cpp:222]: We do, and test for it.

Comment (jonatack) [reply src/wallet/rpcwallet.cpp:222]: Thanks for the quick review! AmountFromValue does raise on negative values and we should already have tests for that.

**Comment by jonatack** [comment]
> in commit [3856f0f](https://github.com/bitcoin/bitcoin/commit/3856f0fdc41fa72e8123f5ba64077d62e8163b4f): Could add test for
> 
>     * 0-fee for `send*` rpcs, and
> 
>     * negative fee for all rpcs?

I think we already have these tests, added in #20305. Will re-verify.

**Comment by jonatack** [comment]
Found some test coverage improvements to make. Updated.

**Comment by MarcoFalke** [comment]
Concept ACK and thanks. Will review when ci is green

**Comment by MarcoFalke** [comment]
ci:

```
�[0m�[0;31mwallet_bumpfee.py                                | ✖ Failed  | 30 s

**Comment by jonatack** [comment]
Yup, not seeing that locally, will update.

**Comment by jonatack** [comment]
Restarted the tasks, they pass, and the failure looks unrelated to this changeset. I can push a patch to loosen the test, though, if wanted.

**Comment by MarcoFalke** [comment]
review ACK a02349dc8250fc22e0cd754b45f379210fbc6e37 did not test 🐟

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK a02349dc8250fc22e0cd754b45f379210fbc6e37 did not test 🐟
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjBDQv+Mogn85V5yxQ2/9YheV57DThEoDMEMllRJi/uc/OQBdQbtq6Kyv2j/6U+
RKE2phLjWZJOlZJk6ip40srFiJ/wuwMKSoGm8/0eWGowncSAVpMRGqKp1ffH27dO
9OMoqyI7dAUbciko/liaQoA+Q1xvLpdg24APW5W97ZVhtVfgh3G0nvzkP7w65sUF
E/WrRvuNbSnG4K2ZNefTGlKLBPUAlH85QUXbQKxcg8GLVn4xNRnWdENaxndGGBXs
m8LNOmshBgT+86X8ODdcArx1BdDQ8wiC9ldrIzaxiGd5CrrBmiNVmBnJ7zbOSQqw
RUq3PtBUymw42l2tgIjRG4bIrMdZRr9to4TiiTKcIkQ1dhDbBSWth/beaiHgvX96
Q8HKPbg/ZWXC6NammYIihG9A5FjlYpzpYpScEGrxBgs7XByaCcRplNrvftSwnlMQ
zL8DjLDAOyGvbgktnnGqH+zg7BiqvGJ1t6qQmeMjt8kMeg7V37vb8STb49O0hKkj
HRIX8M36
=j1gS
-----END PGP SIGNATURE-----
```

python: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory

</details>

File: test/functional/rpc_psbt.py [hunk 2]
```diff
@@ -192,26 +192,27 @@ def run_test(self):
...
+
         self.log.info("Test min fee rate checks with walletcreatefundedpsbt are bypassed, e.g. a fee_rate under 1 sat/vB is allowed")
         res3 = self.nodes[1].walletcreatefundedpsbt(inputs, outputs, 0, {"fee_rate": 0.99999999, "add_inputs": True})
         assert_approx(res3["fee"], 0.00000381, 0.0000001)
         res4 = self.nodes[1].walletcreatefundedpsbt(inputs, outputs, 0, {"feeRate": 0.00000999, "add_inputs": True})
         assert_approx(res4["fee"], 0.00000381, 0.0000001)
 
+        self.log.info("Test min fee rate checks with walletcreatefundedpsbt are bypassed and that funding non-standard 'zero-fee' transactions is valid")
+        for param in ["fee_rate", "feeRate"]:
+            assert_equal(self.nodes[1].walletcreatefundedpsbt(inputs, outputs, 0, {"fee_rate": 0, "add_inputs": True})["fee"], 0)
```
Comment (jonatack) [inline test/functional/rpc_psbt.py:204]:
```suggestion
            assert_equal(self.nodes[1].walletcreatefundedpsbt(inputs, outputs, 0, {param: 0, "add_inputs": True})["fee"], 0)
```

Comment (jonatack) [reply test/functional/rpc_psbt.py]: Updated.

**Comment by jonatack** [comment]
Latest push:
- addition to first commit: fix a test in rpc_psbt.py
- addition to second commit: loosen two tests in wallet_bumpfee.py that may cause spurious CI failures
- addition to third commit: fix "sat vB sat vB" in the bumpfee help

diff ` git diff a02349d 9f08780`:
```diff
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
index a39e8a4e30..dd75e876bc 100644
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -3405,7 +3405,7 @@ static RPCHelpMan bumpfee_helper(std::string method_name)
                     {"fee_rate", RPCArg::Type::AMOUNT, /* default */ "not set, fall back to wallet fee estimation",
                              "\nSpecify a fee rate in " + CURRENCY_ATOM + "/vB instead of relying on the built-in fee estimator.\n"
-                             "Must be at least " + incremental_fee + " " + CURRENCY_ATOM + "/vB higher than the current transaction fee rate.\n"
+                             "Must be at least " + incremental_fee + " higher than the current transaction fee rate.\n"
                              "WARNING: before version 0.21, fee_rate was in " + CURRENCY_UNIT + "/kvB. As of 0.21, fee_rate is in " + CURRENCY_ATOM + "/vB.\n"},
diff --git a/test/functional/rpc_psbt.py b/test/functional/rpc_psbt.py
index 0afa9d7835..5840801b00 100755
--- a/test/functional/rpc_psbt.py
+++ b/test/functional/rpc_psbt.py
@@ -201,7 +201,7 @@ class PSBTTest(BitcoinTestFramework):
 
         self.log.info("Test min fee rate checks with walletcreatefundedpsbt are bypassed and that funding non-standard 'zero-fee' transactions is valid")
         for param in ["fee_rate", "feeRate"]:
-            assert_equal(self.nodes[1].walletcreatefundedpsbt(inputs, outputs, 0, {"fee_rate": 0, "add_inputs": True})["fee"], 0)
+            assert_equal(self.nodes[1].walletcreatefundedpsbt(inputs, outputs, 0, {param: 0, "add_inputs": True})["fee"], 0) 
diff --git a/test/functional/wallet_bumpfee.py b/test/functional/wallet_bumpfee.py
index fab3ab1e27..99c9737258 100755
--- a/test/functional/wallet_bumpfee.py
+++ b/test/functional/wallet_bumpfee.py
@@ -107,12 +107,10 @@ class BumpFeeTest(BitcoinTestFramework):
         # Bumping to just above minrelay should fail to increase the total fee enough.
-        assert_raises_rpc_error(-8, "Insufficient total fee 0.00000141, must be at least 0.00001704 (oldFee 0.00000999 + incrementalFee 0.00000705)",
-            rbf_node.bumpfee, rbfid, {"fee_rate": INSUFFICIENT})
+        assert_raises_rpc_error(-8, "Insufficient total fee 0.00000141", rbf_node.bumpfee, rbfid, {"fee_rate": INSUFFICIENT})
 
         self.log.info("Test invalid fee rate settings")
-        assert_raises_rpc_error(-8, "Insufficient total fee 0.00, must be at least 0.00001704 (oldFee 0.00000999 + incrementalFee 0.00000705)",
-            rbf_node.bumpfee, rbfid, {"fee_rate": 0})
+        assert_raises_rpc_error(-8, "Insufficient total fee 0.00", rbf_node.bumpfee, rbfid, {"fee_rate": 0})
         assert_raises_rpc_error(-4, "Specified or calculated fee 0.141 is too high (cannot be higher than -maxtxfee 0.10",
             rbf_node.bumpfee, rbfid, {"fee_rate": TOO_HIGH})
```

**Comment by MarcoFalke** [comment]
review ACK 9f08780dd7946b63476e9736745131db8e7f4e93  🐾

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 9f08780dd7946b63476e9736745131db8e7f4e93  🐾
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYEACgkQzit1aX5p
pUizQQwAgefxPHhloYKwmNmmM14WEXjPc2tk28ufFSUqXP2V8lgfpOqZ0bGt/ivi
A39fyIC+jgJeACH4WlCC7EPFURgVTDpdUQtHhkCS3Ybl6jEAJ4d5S63Il8nhWLix
LsFFIgZHLRZyGEflQszOWiYUJhtHZrk8aTIX4GfeZbc4dDdIfbx1/LV34yuROYA9
A7ts9zcKuYk8F9SYSNNFTRXU0idA4NJJ3aLE8KBPUf3C/tev8W+cVqccBk6sp0xG
eNeOGdPi1ROjrnIW6QdsiAPfyX7p8pUMeLbz0DT+/jvOMLc7BvX7H0QP1hLzA7HL
hWvVWHzhxfnnwMftIF0aXpJ87M+kX8Jq+pKZXZ2TOKWsrYYhI4eMwQvtnZnoHhIT
XBkUwqZNuesaTADUoRimEhAHrD5lfb30HFOaCY3KziUfmGSLh5/C7WgQmULgN3sa
AvrbWnMYfHe/4MtT67K9w2jAa1HWwgAuYCOLGddR3F3jCCGZ7lu9AYfzaDwUSvze
6IKK1sRg
=NG4i
-----END PGP SIGNATURE-----
```

python: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory

</details>

**Review by promag — COMMENTED** [review]
Code review ACK 9f08780dd7946b63476e9736745131db8e7f4e93.

Maybe makes sense to eventually add `ParseFeeRate(const UniValue&)` .

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -3406,7 +3405,7 @@ static RPCHelpMan bumpfee_helper(std::string method_name)
                     {"conf_target", RPCArg::Type::NUM, /* default */ "wallet -txconfirmtarget", "Confirmation target in blocks\n"},
                     {"fee_rate", RPCArg::Type::AMOUNT, /* default */ "not set, fall back to wallet fee estimation",
                              "\nSpecify a fee rate in " + CURRENCY_ATOM + "/vB instead of relying on the built-in fee estimator.\n"
-                             "Must be at least " + incremental_fee + " " + CURRENCY_ATOM + "/vB higher than the current transaction fee rate.\n"
+                             "Must be at least " + incremental_fee + " higher than the current transaction fee rate.\n"
```
Comment (Xekyo) [inline src/wallet/rpcwallet.cpp:3408]: I am surprised the unit was dropped here. It seems better to be repeat yourself than to give a number without unit to me.

**Review by murchandamus — COMMENTED** [review]
Code review reACK 9f08780dd7946b63476e9736745131db8e7f4e93.

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (Xekyo) [reply src/wallet/rpcwallet.cpp:3408]: Clarified in chat, I didn't understand that the string above already included the unit.

**Comment by jonatack** [comment]
Backported to 0.21 with #20510 (backport tag can be removed).

**Comment by jonatack** [comment]
Backported to 0.21 by #20510.

`Needs backport (0.21)` tag can be removed.

**Comment by MarcoFalke** [comment]
Backported in #20510

**Pull request merged**